Intersect.Client.Framework/Gwen/Control/TreeNode.cs

```diff
@@ -214,31 +214,38 @@ public IEnumerable<TreeNode> SelectedChildren
         public event GwenEventHandler<EventArgs> Collapsed;
 
         /// <summary>
-        ///     Renders the control using specified skin.
+        /// Renders the control using the specified skin.
         /// </summary>
-        /// <param name="skin">Skin to use.</param>
+        /// <param name="skin">The skin to use.</param>
         protected override void Render(Skin.Base skin)
         {
-            var bottom = 0;
-            if (mInnerPanel.Children.Count > 0)
-            {
-                bottom = mInnerPanel.Children.Last().Y + mInnerPanel.Y;
-            }
+            // Calculate the height of the tree node
+            int treeNodeHeight = CalculateTreeNodeHeight();
 
+            // Draw the tree node using the specified skin.
             skin.DrawTreeNode(
-                this, mInnerPanel.IsVisible, IsSelected, mTitle.Height, mTitle.TextRight,
-                (int) (mToggleButton.Y + mToggleButton.Height * 0.5f), bottom, mTreeControl == Parent
+                this, mInnerPanel.IsVisible, IsSelected, treeNodeHeight, mTitle.TextRight,
+                (int)(mToggleButton.Y + mToggleButton.Height * 0.5f), mInnerPanel.Bottom, mTreeControl == Parent
             ); // IsRoot
 
-            //[halfofastaple] HACK - The treenodes are taking two passes until their height is set correctly,
-            //  this means that the height is being read incorrectly by the parent, causing
-            //  the TreeNode bug where nodes get hidden when expanding and collapsing.
-            //  The hack is to constantly invalide TreeNodes, which isn't bad, but there is
-            //  definitely a better solution (possibly: Make it set the height from childmost
-            //  first and work it's way up?) that invalidates and draws properly in 1 loop.
+            // Invalidate the tree node.
             this.Invalidate();
         }
 
+        /// <summary>
+        /// Calculates the height of tree node.
+        /// </summary>
+        private int CalculateTreeNodeHeight()
+        {
+            int height = mTitle.Height;
+            if (mInnerPanel.Children.Count > 0)
```

Comment:
When Count is 0, what is Height?

If Height is 0 when Count is 0, just make this `=> mTitle.Height + mInnerPanel.Height`

Reply:
Alright ⛑️ ~
So far, this gave me better and essentially the best outcome, this also seems to have completly fixed the node tree issues with folders / collapsable nodes:

```csharp
        /// <summary>
        /// Calculates the height of tree node.
        /// </summary>
        private int CalculateTreeNodeHeight()
        {
            var height = mTitle.Height;

            if (mInnerPanel.Children.Count > 0)
            {
                height = mInnerPanel.Children.Last().Y + height;
            }
            else if (height == 0)
            {
                height = mInnerPanel.Height;
            }

            return height;
        }
```

![imagen](https://user-images.githubusercontent.com/17498701/222311209-23d7738c-7046-4e52-b3aa-1b04d51e27df.png) ![imagen](https://user-images.githubusercontent.com/17498701/222311231-e508b9da-d893-4600-a573-4bacf5709f1f.png) ![imagen](https://user-images.githubusercontent.com/17498701/222311246-e004cd79-0ad9-405d-8366-045bf2472082.png) ![imagen](https://user-images.githubusercontent.com/17498701/222311266-d9714116-7cd9-414a-aa47-2fbb112a29aa.png) ![imagen](https://user-images.githubusercontent.com/17498701/222311287-45d91525-82a1-4d4c-87f6-48914ae64108.png) ![imagen](https://user-images.githubusercontent.com/17498701/222311314-22beb1fc-4562-4622-b79d-01032566baa8.png)